### PR TITLE
EG-1134 date input

### PR DESCRIPTION
**Pull request by LesWilson**
Merged in current master changes

### CODE REVIEW

File: app/services/PageBuilder.scala [hunk 1]
```diff
@@ -133,6 +133,8 @@ class PageBuilder extends ProcessPopulation {
           f(page.id, page.url, text.langs(0))
```
Comment (woollett-r) [inline app/services/PageBuilder.scala:133]: The methods fromPageDetails and pagesWithValidation are used only in the test classes which makes me wonder if they should be relocated into the test classes.

Comment (cmrhmrc) [reply app/services/PageBuilder.scala:133]: Richard, fromPageDetails and pagesWithValidation are both used in EG and this code is shared between both the viewer and EG

File: app/controllers/GuidanceController.scala [hunk 2]
```diff
@@ -45,13 +46,14 @@ class GuidanceController @Inject() (
...
     formProvider: SubmittedAnswerFormProvider,
     service: GuidanceService,
     mcc: MessagesControllerComponents
 ) extends FrontendController(mcc)
   with SessionFrontendController {
 
-  val logger = Logger(getClass)
+  val logger: Logger = Logger(getClass)
 
   def getPage(processCode: String, path: String, p: Option[String]): Action[AnyContent] = sessionIdAction.async { implicit request =>
```
Comment (woollett-r) [inline app/controllers/GuidanceController.scala:58]: This method is exceeding the cyclomatic complexity threshold. I'd be tempted to move the code handling a successful data retrieval (lines 62 to 76 to a private method). I have done something similar for the submission code in EG-1137-Exp.

File: app/views/components/input_date.scala.html [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+      @components.h1_heading_with_label(H1(input.text), name + "-0")
+      @{input.hint.map(hintText => <span id="input-hint" class="govuk-hint">{components.render_text(hintText)}</span>)}
+      @components.error_message(input.errorMsgs)
+      @components.input_date_field(input, name, form)
+
+  } else {
+    @components.h1_heading(H1(input.text))
+    @components.render_components(input.body)
+
+    <fieldset class="govuk-fieldset" @optionalAria()>
```
Comment (woollett-r) [inline app/views/components/input_date.scala.html:42]: Why is the fieldset element only used when there are additional components on the page?

Comment (LesWilson) [reply app/views/components/input_date.scala.html:42]: That was what was thrown up in the accessibility review to remove it from single input - but matches the GDS style guide for date input - a fieldset groups fields which are related

File: app/views/components/input_date_field.scala.html [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+@(input: models.ui.DateInput, name: String, form: Form[_])(implicit request: Request[_], messages: Messages, ctx: PageContext)
+
+@optionalAria() = @{
+  Seq(input.hint.map(_ => s"$name-item-hint"), input.errorMsgs.headOption.map(err => s"${err.id}-error")).flatten match {
+    case Nil => ""
+    case l => ariaDescribedBy("input-hint")
+  }
+}
+<div class="govuk-date-input" id="date-input">
```
Comment (woollett-r) [inline app/views/components/input_date_field.scala.html:25]: The id here is not going to be unique if there are multiple dates on a form. Further, if there are multiple components on a page a label is created for the id "name-0", should this be the id?

File: app/views/components/input_date_field.scala.html [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    <div class="govuk-date-input__item">
+        <div class="govuk-form-group">
+            <label class="govuk-label govuk-date-input__label" for="day">
+              @messages("label.day")
+            </label>
+            <input class="govuk-input govuk-date-input__input govuk-input--width-2" id="day" name="day" value="@form("day").value" type="text" pattern="[0-9]*" inputmode="numeric">
```
Comment (woollett-r) [inline app/views/components/input_date_field.scala.html:31]: Would it be better to use @name-day here?

File: app/views/components/input_date_field.scala.html [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    </div>
+    <div class="govuk-date-input__item">
+        <div class="govuk-form-group">
+            <label class="govuk-label govuk-date-input__label" for="year">
+               @messages("label.year")
+            </label>
+            <input class="govuk-input govuk-date-input__input govuk-input--width-4" id="year" name="year" value="@form("year").value" type="text" pattern="[0-9]*" inputmode="numeric"></div>
+    </div>
+</div>
+@*<div class="govuk-form-group">*@
```
Comment (woollett-r) [inline app/views/components/input_date_field.scala.html:49]: If these lines are commented out they should be removed.

File: app/views/components/input_date.scala.html (hunk 3 above)
Comment (woollett-r) [reply app/views/components/input_date.scala.html:42]: Ok

File: app/services/PageBuilder.scala (hunk 1 above)
Comment (woollett-r) [reply app/services/PageBuilder.scala:133]: Ok, I can one usage in external-guidance.

File: app/controllers/GuidanceController.scala (hunk 2 above)
Comment (LesWilson) [reply app/controllers/GuidanceController.scala:58]: will take a look

File: app/views/components/input_date_field.scala.html (hunk 4 above)
Comment (LesWilson) [reply app/views/components/input_date_field.scala.html:25]: As this assumes one input per page, adding -0 at end of id doesn't really make much sense.  The date fields are called day, month and year, so would have to change if we have multiple dates on a page - but this woud be the least of the changes compared to controller side of things.

File: app/views/components/input_date_field.scala.html (hunk 5 above)
Comment (LesWilson) [reply app/views/components/input_date_field.scala.html:31]: see above, not sure why we need the @name as we will have a specific date handler in back end

File: app/views/components/input_date_field.scala.html (hunk 6 above)
Comment (LesWilson) [reply app/views/components/input_date_field.scala.html:49]: yes, will sort out

File: app/views/components/input_date_field.scala.html (hunk 4 above)
Comment (woollett-r) [reply app/views/components/input_date_field.scala.html:25]: Ok

File: app/views/components/input_date_field.scala.html (hunk 5 above)
Comment (woollett-r) [reply app/views/components/input_date_field.scala.html:31]: Yes, it all depends on how many input components we have on a page. If it is only one this is ok.

File: test/controllers/GuidanceControllerSpec.scala [hunk 7]
```diff
@@ -27,15 +28,16 @@ import play.api.test.Helpers._
 import uk.gov.hmrc.play.bootstrap.tools.Stubs.stubMessagesControllerComponents
 import uk.gov.hmrc.http.SessionKeys
 import forms.SubmittedAnswerFormProvider
-import models.{PageEvaluationContext, PageContext}
-import models.ocelot.{Process, ProcessJson, Phrase,KeyedStanza, Page => OcelotPage}
-import models.ocelot.stanzas.{Question => OcelotQuestion,_}
-import models.ui.{CurrencyInput => UiCurrencyInput,_}
+import models.{PageContext, PageEvaluationContext}
+import models.ocelot.{KeyedStanza, Phrase, Process, ProcessJson, Page => OcelotPage}
```
Comment (woollett-r) [inline test/controllers/GuidanceControllerSpec.scala:32]: Rearranging the order to alphabetic order; that's the kind of OCD update I do!

Comment (LesWilson) [reply test/controllers/GuidanceControllerSpec.scala:32]: Thats the sort of OCD update Intellij does :)

File: test/views/components/InputSpec.scala [hunk 8]
```diff
@@ -153,4 +153,95 @@ class InputSpec extends WordSpec with Matchers with GuiceOneAppPerSuite {
       }
    }
   }
+
+  "English Date Input component" must {
```
Comment (woollett-r) [inline test/views/components/InputSpec.scala:157]: For completeness I'd add a test for the existence of a fieldset when there are additional components on the page.

File: app/views/components/h1_heading_with_fieldset.scala.html [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *@
+
+@import ui.H1
+
+@( h1: H1 )( implicit messages: Messages, ctx: PageContext)
+<legend class="govuk-fieldset__legend govuk-fieldset__legend--xl">
```
Comment (woollett-r) [inline app/views/components/h1_heading_with_fieldset.scala.html:20]: xl seems on the large size. What is your reasoning?

Comment (woollett-r) [reply app/views/components/h1_heading_with_fieldset.scala.html:20]: I see question uses xl so I guess it is consistent.

Comment (LesWilson) [reply app/views/components/h1_heading_with_fieldset.scala.html:20]: It matched the input with label component

Comment (cmrhmrc) [reply app/views/components/h1_heading_with_fieldset.scala.html:20]: have a look at h1_heading.scala.html. There is a setting on the page which determines if the reduced heading sizes should be used. something like class="@if(ctx.page.reduceHeadings){govuk-heading-l} else {govuk-heading-xl}"

Comment (cmrhmrc) [reply app/views/components/h1_heading_with_fieldset.scala.html:20]: also applies to h1_heading_with_label.scala.html

**Pull request merged**